File: go/tree/nary_tree_preorder_traversal.go
package main

import "container/list"

//N叉树的前序遍历

//使用递归
func preorder(root *Node) []int {
	var res []int
	npreHelper(root, &res)
	return res
}

func npreHelper(node *Node, res *[]int) {
	if node == nil {
		return
	}

	*res = append(*res, node.Val)
	for _, child := range node.Children {
		npreHelper(child, res)
	}
}

//使用迭代
func preorder2(root *Node) []int {
	var res []int
	if root == nil {
		return res
	}

	stack := list.New()
	stack.PushFront(root)
	for stack.Len() > 0 {
		e := stack.Front()
		stack.Remove(e)
		node := e.Value.(*Node)
		res = append(res, node.Val)
		for i := len(node.Children) - 1; i >= 0; i-- {
			stack.PushFront(node.Children[i])
		}
	}
	return res
}

type Node struct {
	Val      int
	Children []*Node
}

File: java/datastructure/tree/leetcode/NaryTreePreorderTraversal.java
package datastructure.tree.leetcode;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Stack;

/**
 * @author roseduan
 * @time 2020/9/19 10:52 上午
 * @description N叉树的前序遍历
 */
public class NaryTreePreorderTraversal {

    public static void main(String[] args) {
        String s = "ANNS";
        s.toLowerCase();
    }

    /**
     * 递归
     */
    private final List<Integer> res = new ArrayList<>();
    public List<Integer> preorder(Node root) {
        if (root == null) {
            return res;
        }

        res.add(root.val);
        List<Node> children = root.children;
        if (children != null) {
            for (Node node : children) {
                preorder(node);
            }
        }
        return res;
    }

    /**
     * 迭代
     */
    public List<Integer> preorder2(Node root) {
        List<Integer> list = new ArrayList<>();
        if (root == null) {
            return list;
        }

        Stack<Node> stack = new Stack<>();
        stack.push(root);
        while (!stack.empty()) {
            Node node = stack.pop();
            list.add(node.val);
            List<Node> children = node.children;
            if (children != null) {
                Collections.reverse(children);
                for (Node child : children) {
                    stack.push(child);
                }
            }
        }
        return res;
    }

    class Node {
        public int val;
        public List<Node> children;
        public Node() {}
        public Node(int _val) {
            val = _val;
        }
        public Node(int _val, List<Node> _children) {
            val = _val;
            children = _children;
        }
    };
}
